sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala

```diff
@@ -100,27 +101,128 @@ class SpillableColumnarBatchImpl (
    */
   override def numRows(): Int = rowCount
 
-  private def withRapidsBuffer[T](fn: RapidsBuffer => T): T = {
-    withResource(RapidsBufferCatalog.acquireBuffer(handle)) { rapidsBuffer =>
-      fn(rapidsBuffer)
+  override lazy val sizeInBytes: Long = handle.sizeInBytes
+
+  /**
+   * Set a new spill priority.
+   */
+  override def setSpillPriority(priority: Long): Unit = {
+    // TODO: handle.setSpillPriority(priority)
+  }
+
+  override def getColumnarBatch(): ColumnarBatch = {
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    handle.materialize(sparkTypes)
+  }
+
+  override def incRefCount(): SpillableColumnarBatch = {
+    if (refCount <= 0) {
+      throw new IllegalStateException("Use after free on SpillableColumnarBatchImpl")
     }
+    refCount += 1
+    this
   }
 
-  override lazy val sizeInBytes: Long =
-    withRapidsBuffer(_.memoryUsedBytes)
+  /**
+   * Remove the `ColumnarBatch` from the cache.
+   */
+  override def close(): Unit = {
+    refCount -= 1
+    if (refCount == 0) {
+      // closing my reference
+      handle.close()
+    }
+    // TODO this is causing problems so we need to look into this
+    //  https://github.com/NVIDIA/spark-rapids/issues/10161
+    //else if (refCount < 0) {
+    //  throw new IllegalStateException("Double free on SpillableColumnarBatchImpl")
+    //}
+  }
+
+  override def toString: String =
+    s"SCB $handle $rowCount ${sparkTypes.toList} $refCount"
+}
+
+class SpillableCompressedColumnarBatchImpl(
+    handle: SpillableCompressedColumnarBatchHandle, rowCount: Int)
+  extends SpillableColumnarBatch {
+
+  private var refCount = 1
+
+  /**
+   * The number of rows stored in this batch.
+   */
+  override def numRows(): Int = rowCount
+
+  override lazy val sizeInBytes: Long = handle.compressedSizeInBytes
 
   /**
    * Set a new spill priority.
    */
   override def setSpillPriority(priority: Long): Unit = {
-    handle.setSpillPriority(priority)
+    // TODO: handle.setSpillPriority(priority)
   }
 
   override def getColumnarBatch(): ColumnarBatch = {
-    withRapidsBuffer { rapidsBuffer =>
-      GpuSemaphore.acquireIfNecessary(TaskContext.get())
-      rapidsBuffer.getColumnarBatch(sparkTypes)
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    handle.materialize
+  }
+
+  override def incRefCount(): SpillableColumnarBatch = {
+    if (refCount <= 0) {
+      throw new IllegalStateException("Use after free on SpillableColumnarBatchImpl")
+    }
+    refCount += 1
+    this
+  }
+
+  /**
+   * Remove the `ColumnarBatch` from the cache.
+   */
+  override def close(): Unit = {
+    refCount -= 1
+    if (refCount == 0) {
+      // closing my reference
+      handle.close()
     }
+    // TODO this is causing problems so we need to look into this
```

Comment:
again this PR is big enough. I think we can do this in a separate PR.

Reply:
Okay Sorry it looks like these comments were here already. This is just copied from previous code.

Reply:
yeah let me punt on these comments for now.